### PR TITLE
Fix params corner cases

### PR DESCRIPTION
**Pull request by dividedmind**
Fixes handling of parameters in two corner cases:
- missing required argument,
- method called with self=None.

The latter caused an error when I was trying to map saleor ([`saleor/graphql/page/tests/test_page_types_sorting_and_filtering.py::test_sort_page_types_by_slug`](https://github.com/mirumee/saleor/blob/master/saleor/graphql/page/tests/test_page_types_sorting_and_filtering.py#L70), for reference).

### CODE REVIEW

File: appmap/_implementation/event.py [hunk 1]
```diff
@@ -183,9 +183,21 @@ def set_params(params, instance, args, kwargs):
         # dict, respectively. It operates on them as collections, so
         # it doesn't unpack them.
         ret = []
-        if instance is not None:
-            ret.append(params[0].to_dict(instance))
-            params = params[1:]
+
+        # HACK: this is to detect cases when this is a method, yet the self
+        # parameter is None. Unfortunately wrapt tries to be too smart
```
[the remark this answers is not in the document]
Comment (dividedmind) [reply appmap/_implementation/event.py:188]: Not necessarily; there's nothing invalid about calling method with self=None. Look at saleor code I've linked to. Or consider something like this (off the top of my head):
```
class Filter:
  def skip(self, value):
    return value

  def filter_one_way(self, value):
    # do something with self

  def filter_another_way(self, value):
    # do something with self

def maybe_filter(value, filter=None, filter_method=Filter.skip)
  return filter_method(filter, value)

maybe_filter('foo')
maybe_filter('bar', get_some_filter(), Filter.filter_another_way)
```

File: appmap/_implementation/event.py [hunk 2]
```diff
@@ -183,9 +183,21 @@ def set_params(params, instance, args, kwargs):
...
+        # by separating the instance argument, and as a consequence
+        # we can't differentiate here between a function and a method
+        # bound to None. Thus we rely on 'self' being the conventional
+        # name for the self argument. This is usually correct but
+        # theoretically could be wrong with code that's off-style.
+        takes_self = params and params[0].name == 'self'
```
[the remark this answers is not in the document]
Comment (dividedmind) [reply appmap/_implementation/event.py:194]: Theoretically, but I haven't seen that in the wild and I can't imagine how this could be useful; since this is a bit of a hack I decided to limit it to what was actually needed.

**Review by apotterri — APPROVED** [review]
Ok, this makes sense to me now. Thanks.
